crates/client/mempool/src/inner.rs

```
@@ -0,0 +1,536 @@
+            if let AccountTransaction::DeployAccount(tx) = &contract.0.tx {
+                debug_assert!(deployed_contracts.remove(&tx.contract_address))
+            };
+        }
+        debug_assert!(deployed_contracts.is_empty());
+    }
+
+    /// When `force` is `true`, this function should never return any error.
+    pub fn insert_tx(&mut self, mempool_tx: MempoolTransaction, force: bool) -> Result<(), TxInsersionError> {
+        // Get the nonce chain for the contract
+
+        let contract_addr = mempool_tx.contract_address();
+        let arrived_at = mempool_tx.arrived_at;
+
+        let deployed_contract_address =
+            if let AccountTransaction::DeployAccount(tx) = &mempool_tx.tx { Some(tx.contract_address) } else { None };
+
+        if let Some(contract_address) = &deployed_contract_address {
+            if !self.deployed_contracts.insert(*contract_address) && !force {
+                return Err(TxInsersionError::AccountAlreadyDeployed);
+            }
+        }
+
+        match self.nonce_chains.entry(contract_addr) {
+            hash_map::Entry::Occupied(mut entry) => {
+                // Handle nonce collision.
+                let position = match entry.get_mut().insert(mempool_tx, force) {
+                    Ok(position) => position,
+                    Err(_nonce_collision) => {
+                        if force {
+                            panic!("Force add should never error")
+                        }
+                        // Rollback the prior mutation.
+                        if let Some(contract_address) = &deployed_contract_address {
+                            if !self.deployed_contracts.remove(contract_address) {
+                                return Err(TxInsersionError::AccountAlreadyDeployed);
+                            }
+                        }
+
+                        return Err(TxInsersionError::NonceConflict);
```

should we return the same error thrown by `insert`?

---

sure, i wanted to be extra-verbose here in this file since it's more critical